Add URLPatternComponent class

jelee53 opened the pull request:
#### d3b30382f58c51f51d0762fde51c10e011be56d6
<pre>
Add URLPatternComponent class
<a href="https://bugs.webkit.org/show_bug.cgi?id=283073">https://bugs.webkit.org/show_bug.cgi?id=283073</a>
<a href="https://rdar.apple.com/139833314">rdar://139833314</a>

Reviewed by Sihui Liu.

Add preliminary support for URLPatternComponent, see <a href="https://urlpattern.spec.whatwg.org/#component">https://urlpattern.spec.whatwg.org/#component</a>

* LayoutTests/fast/urlpattern/urlpattern-component-expected.txt:
* LayoutTests/fast/urlpattern/urlpattern-component.html:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WTF/wtf/URL.cpp:
(WTF::URL::isIPv6Address):
(WTF::isIPv6Address): Deleted.
* Source/WTF/wtf/URL.h:
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::CodePoints::codePointAt const):
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::processBaseURLString):
(WebCore::URLPattern::create):
(WebCore::URLPattern::compileAllComponents):
(WebCore::escapePatternStringForCharacters): Deleted.
(WebCore::escapePatternString): Deleted.
(WebCore::URLPattern::URLPattern): Deleted.
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Modules/url-pattern/URLPattern.idl:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizeOpaquePathname):
(WebCore::canonicalizePathname):
(WebCore::callEncodingCallback):
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.h:
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp: Added.
(WebCore::URLPatternUtilities::URLPatternComponent::URLPatternComponent):
(WebCore::URLPatternUtilities::URLPatternComponent::compile):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.h: Copied from Source/WebCore/Modules/url-pattern/URLPatternCanonical.h.
(WebCore::URLPatternUtilities::URLPatternComponent::patternString const):
* Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
(WebCore::URLPatternUtilities::URLPatternParser::performParse):
(WebCore::URLPatternUtilities::URLPatternParser::addPart):
(WebCore::URLPatternUtilities::URLPatternParser::parse):
(WebCore::URLPatternUtilities::escapeRegexString):
(WebCore::URLPatternUtilities::convertModifierToString):
(WebCore::URLPatternUtilities::generateRegexAndNameList):
(WebCore::URLPatternUtilities::generatePatternString):
(WebCore::URLPatternUtilities::escapePatternStringForCharacters):
(WebCore::URLPatternUtilities::escapePatternString):
(WebCore::URLPatternUtilities::isValidNameCodepoint):
(): Deleted.
* Source/WebCore/Modules/url-pattern/URLPatternParser.h:
(WebCore::URLPatternUtilities::URLPatternParser::takePartList):
(WebCore::URLPatternUtilities::URLPatternParser::getPartList const): Deleted.
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
(WebCore::URLPatternUtilities::Tokenizer::tokenize):
(WebCore::URLPatternUtilities::isValidNameCodepoint): Deleted.
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/286965@main">https://commits.webkit.org/286965@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9dea3cd60bdaa106334c22ddd5f09e8728f94c8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77654 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56687 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30576 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/82279 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28932 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79771 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65843 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4987 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/60857 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18821 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80720 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50815 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66644 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41131 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48205 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24142 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27263 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/70842 "Found 4 new JSC stress test failures: stress/attribute-custom-accessor.js.ftl-eager, stress/get-by-val-hoist-above-structure.js.ftl-eager-no-cjit-b3o1, stress/get-by-val-hoist-above-structure.js.ftl-no-cjit-no-put-stack-validate, wasm.yaml/wasm/v8/regress/regress-824681.js.wasm-collect-continuously (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69319 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24486 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83657 "Built successfully") | 
| [❌ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/76931 "Found 2 new JSC stress test failures: stress/attribute-custom-accessor.js.ftl-eager, wasm.yaml/wasm/modules/wasm-imports-js-exports.js.wasm-aggressive-inline (failure)") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5034 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3422 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69084 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5191 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66614 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68336 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12340 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10444 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/99192 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12037 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4981 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/7790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21679 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4996 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8434 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6759 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -222,16 +198,23 @@ static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStr
 {
     UNUSED_PARAM(baseURL);
 
+    RefPtr<VM> vm = &(context.vm());
```
`Ref vm = context.vm();` but more importantly, this is not needed. vm is not used at all in this function. Let's not add code until we actually need it.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -222,16 +198,23 @@ static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStr
     UNUSED_PARAM(baseURL);
 
+    RefPtr<VM> vm = &(context.vm());
+    JSLockHolder lock(*vm);
```
can omit the star with the change above. But also, not needed at the moment.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -222,16 +198,23 @@ static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStr
-ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&& input, URLPatternOptions&&)
+ExceptionOr<Ref<URLPattern>> URLPattern::create(ScriptExecutionContext& context, std::optional<URLPatternInput>&& input, URLPatternOptions&& options)
 {
+    RefPtr<VM> vm = &(context.vm());
```
ditto.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -268,20 +250,14 @@ ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&&
-    // FIXME: Implement compiling the component for each URLPattern member field to replace this current placeholder code which reroutes processed URLPatternInit object to URLPattern result.
-    return adoptRef(*new URLPattern(WTFMove(processedInit)));
-}
+    auto result = adoptRef(*new URLPattern());
```
`auto` -> `Ref`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -300,4 +276,71 @@ ExceptionOr<std::optional<URLPatternResult>> URLPattern::exec(std::optional<URLP
+// https://urlpattern.spec.whatwg.org/#hostname-pattern-is-an-ipv6-address
+static bool isIPv6Address(StringView input)
+{
+    Vector<UChar32> inputCodepoints(input.length());
```
There must be a way to write this function without having to copy input into a new buffer. This is very inefficient.

Also, there's already a likely more efficient `isIPv6Address()` in URL.cpp that we can likely reuse.

jelee53 replied:
Yes, there is. Had to refactor URL.cpp a little bit to use it though.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -0,0 +1,79 @@
+
+const String& URLPatternComponent::getPatternString() const
+{
+    return WTFMove(m_patternString);
```
This definitely should NOT use WTFMove(). Also, it can be inlined in the header.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -0,0 +1,79 @@
+
+    bool hasRegexGroups = false;
+
+    for (const Part& part : partList) {
```
`auto&`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -0,0 +1,79 @@
+
+    for (const Part& part : partList) {
+        if (part.type == PartType::Regexp)
+            hasRegexGroups = true;
```
You need to `break;` after you find a match.

cdumez replied:
An easy way to avoid making mistakes like this is to reuse existing algorithms like `Vector::containsIf()`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.h:
```diff
@@ -0,0 +1,52 @@
+
+private:
+    String m_patternString;
+    JSC::RegExp* m_regularExpression;
```
This likely needs to be stored in a JSC::Strong smart pointer to keep it alive.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
 }
 
+// https://urlpattern.spec.whatwg.org/#convert-a-modifier-to-a-string
+String convertModifierToString(Modifier modifier)
```
Should return an `ASCIILiteral` to avoid unnecessary String allocations.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+        return "?"_s;
+    case Modifier::OneOrMore:
+        return "+"_s;
+    default:
```
Do we need the default case here?

jelee53 replied:
Yes because this doesn't account for all the values, eg Modifier::None. In same cases, we want to return nothing.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+std::pair<String, Vector<String>> generateRegexAndNameList(const Vector<Part>& partList, const URLPatternStringOptions& options)
+{
+    StringBuilder result;
+    result.append("^"_s);
```
`result.append('^');`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+
+    Vector<String> nameList;
+
+    for (const Part& part : partList) {
```
`auto&`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+        StringView regexpValue = part.value;
+
+        if (part.type == PartType::SegmentWildcard)
+            regexpValue = generateSegmentWildcardRegexp(options);
```
This cannot be safe. This is likely returning a newly constructed String, then you're constructing a StringView from it, then the temporary String gets destroyed and your StringView points to stale memory. Clear security bug.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+    StringBuilder result;
+
+    for (size_t index = 0; index < partList.size(); ++index) {
+        const Part& part = partList[index];
```
`auto&`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+            continue;
+        }
+
+        bool customName = !part.name.isEmpty() && !isASCIIDigit(part.name[0]);
```
boolean variables need a prefix per WebKit coding style. E.g. "isCustomName"

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+            if (part.modifier == Modifier::None)
+                result.append(escapeRegexString(part.value));
+            else
+                result.append(makeString("(?:"_s, escapeRegexString(part.value), ")"_s, convertModifierToString(part.modifier)));
```
It is inefficient to construct a temporary string just to append to a StringBuilder. I think you can write it more efficiently like so:
```
result.append("(?:"_s, escapeRegexString(part.value), ")"_s, convertModifierToString(part.modifier));
```
If that doesn't work, there's always:
```
result.append("(?:"_s);
result.append(escapeRegexString(part.value));
...
```

cdumez replied:
The same comment likely applies in other places.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+            result.append("?"_s);
+    }
+
+    result.append("$"_s);
```
`'$'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+            ")"_s);
+
+        if (part.modifier == Modifier::ZeroOrMore)
+            result.append("?"_s);
```
`'?'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+
+                continue;
+            }
+            result.append(makeString("{"_s, escapePatternString(part.value), "}"_s, convertModifierToString(part.modifier)));
```
Should not construct a temporary string.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+        result.append(escapePatternString(part.prefix));
+
+        if (customName)
+            result.append(makeString(":"_s, part.name));
```
Should not create a temp string.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+        if (part.type == PartType::Regexp)
+            result.append("("_s, part.value, ")"_s);
+        else if (part.type == PartType::SegmentWildcard && !customName)
+            result.append(makeString("("_s, generateSegmentWildcardRegexp(options), ")"_s));
```
Should not create a temp string.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+            if (!customName
+                && (!previousPart || previousPart->type == PartType::FixedText || previousPart->modifier != Modifier::None
+                    || needsGrouping || !part.prefix.isEmpty()))
+                result.append("*"_s);
```
`'*'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+                    || needsGrouping || !part.prefix.isEmpty()))
+                result.append("*"_s);
+            else
+                result.append(makeString("("_s, ".*"_s,  ")"_s));
```
Seems like the least efficient way to write:
```
result.append("(.*)"_s);
```

jelee53 replied:
Yup! My bad, the spec confused me a little bit with this. This has been fixed. 
<img width="579" alt="Screenshot 2024-11-15 at 6 09 03 PM" src="https://github.com/user-attachments/assets/2be45bff-3646-4705-84ab-81a9c341e4d8">

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+        result.append(escapePatternString(part.suffix));
+
+        if (needsGrouping)
+            result.append("}"_s);
```
`'}'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+        }
+
+        if (part.type == PartType::SegmentWildcard && customName && !part.suffix.isEmpty() && isValidNameCodepoint(*StringView(part.suffix).codePoints().begin(), true))
+            result.append("\\"_s);
```
`'\\'`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,219 @@ String escapeRegexString(StringView input)
+
+
+// https://urlpattern.spec.whatwg.org/#is-a-valid-name-code-point
+bool isValidNameCodepoint(UChar codepoint, bool first)
```
The second parameter should be a enum class like:
```
enum class IsFirst : bool { No, Yes };
```
It would make the call sites a lot more readable.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -0,0 +1,79 @@
+        return maybePartList.releaseException();
+    Vector<Part> partList = maybePartList.releaseReturnValue();
+
+    auto regexAndNameList = generateRegexAndNameList(partList, options);
```
I think this would be better as:
```
auto [regularExpressionString, nameList] = generateRegexAndNameList(partList, options);
```
It would avoid having a "moved-out" `regexAndNameList` in scope for the rest of the function, which is super error-prone.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -300,4 +270,50 @@ ExceptionOr<std::optional<URLPatternResult>> URLPattern::exec(std::optional<URLP
+    m_protocolComponent = maybeProtocolComponent.releaseReturnValue();
+
+    auto maybeUsernameComponent = URLPatternUtilities::URLPatternComponent::compile(context.vm(), processedInit.username, EncodingCallbackType::Username, URLPatternUtilities::URLPatternStringOptions { });
+    if (maybeProtocolComponent.hasException())
```
Should be `maybeUsernameComponent`

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -300,4 +270,50 @@ ExceptionOr<std::optional<URLPatternResult>> URLPattern::exec(std::optional<URLP
+
+    auto maybeUsernameComponent = URLPatternUtilities::URLPatternComponent::compile(context.vm(), processedInit.username, EncodingCallbackType::Username, URLPatternUtilities::URLPatternStringOptions { });
+    if (maybeProtocolComponent.hasException())
+        return maybeProtocolComponent.releaseException();
```
ditto.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.h:
```diff
@@ -0,0 +1,52 @@
+    JSC::Strong<JSC::RegExp> m_regularExpression;
+    Vector<String> m_groupNameList;
+    bool m_hasRegexpGroups;
+
```
Extra blank line

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.h:
```diff
@@ -0,0 +1,52 @@
+    static ExceptionOr<URLPatternComponent> compile(Ref<JSC::VM>, StringView, EncodingCallbackType, const URLPatternStringOptions&);
+    const String& patternString() const { return m_patternString; }
+
+    URLPatternComponent(String&&,  JSC::Strong<JSC::RegExp>, Vector<String>&&, bool);
```
Extra space; also should be `JSC::Strong<JSC::RegExp>&&`.
Also, do we expect `URLPatternComponent` to be constructed from other means than `compile` function? If not, the constructor should be private.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,214 @@ String escapeRegexString(StringView input)
+    case Modifier::OneOrMore:
+        return "+"_s;
+    default:
+            return { };
```
Extra indentation

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,214 @@ String escapeRegexString(StringView input)
+            && part.type == PartType::SegmentWildcard && part.modifier == Modifier::None
+            && nextPart && nextPart->prefix.isEmpty() && nextPart->suffix.isEmpty()) {
+            if (nextPart->type == PartType::FixedText) {
+                // FIXME: Figure out if should be true or  false.
```
Need some clarification for the FIXME comment

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -330,5 +329,214 @@ String escapeRegexString(StringView input)
+                // FIXME: Figure out if should be true or  false.
+                needsGrouping = isValidNameCodepoint(*StringView(nextPart->value).codePoints().begin(), IsFirst::Yes);
+            } else
+                needsGrouping = isASCIIDigit(nextPart->name[0]);
```
Is `name` guaranteed to be non-empty? Seems should be `!nextPart->name.isEmpty() && isASCIIDigit(nextPart->name[0])`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -280,19 +280,18 @@ bool URLPatternParser::isDuplicateName(StringView name) const
     tokenParser.performParse(options);
 
-    return tokenParser.getPartList();
+    return WTFMove(tokenParser.m_partList);
```
We could convert `getPartList()` to
`Vector<Part> takePartList() { return std::exchange(m_partList, { ); }` and call `return tokenParser.takePartList();` here.
Also `takePartList()` should be private.

cdumez commented on LayoutTests/fast/urlpattern/urlpattern-component-expected.txt:
```diff
@@ -2796,7 +2682,7 @@ Testing: {"pattern":[{"username":"caf%C3%A9"}],"inputs":[{"username":"cafÃ©"}]
 PASS urlPatternComponent is "*"
 Component: username
-PASS urlPatternComponent is "caf%C3%A9"
+FAIL urlPatternComponent should be caf%C3%A9. Was café.
```
Shouldn't I be concerned about this?

cdumez replied:
ditto for other failures below.

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -322,13 +325,220 @@ static String escapeRegexStringForCharacters(std::span<const CharacterType> char
+// https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
+String escapePatternString(StringView input)
+{
+    // FIXME: Add check to see that input only contains ASCII
```
Do you mean "ensure that input only contains ASCII based on spec"?
I think we need to add that "after the parser (or tokenizer) knows to filter non-ASCII input".

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
```diff
@@ -322,13 +325,220 @@ static String escapeRegexStringForCharacters(std::span<const CharacterType> char
+// https://urlpattern.spec.whatwg.org/#is-a-valid-name-code-point
+bool isValidNameCodepoint(UChar codepoint, URLPatternUtilities::IsFirst first)
+{
+    if (static_cast<bool>(first))
```
Just compare first with `URLPatternUtilities::IsFirst` values and avoid `static_cast`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
```diff
@@ -160,7 +152,7 @@ ExceptionOr<Vector<Token>> Tokenizer::tokenize()
                 seekNextCodePoint(namePosition);
 
-                bool isValidCodepoint = isValidNameCodepoint(m_codepoint, namePosition == nameStart);
+                bool isValidCodepoint = isValidNameCodepoint(m_codepoint, static_cast<IsFirst>(namePosition == nameStart));
```
`namePosition == nameStart ? IsFirst::Yes : IsFirst::No`

szewai commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.h:
```diff
@@ -0,0 +1,51 @@
+    String m_patternString;
+    JSC::Strong<JSC::RegExp> m_regularExpression;
+    Vector<String> m_groupNameList;
+    bool m_hasRegexpGroups;
```
Needs initialization for default constructor

cdumez commented on Source/WebCore/Modules/url-pattern/URLPattern.cpp:
```diff
@@ -268,20 +245,13 @@ ExceptionOr<Ref<URLPattern>> URLPattern::create(std::optional<URLPatternInput>&&
-    // FIXME: Implement compiling the component for each URLPattern member field to replace this current placeholder code which reroutes processed URLPatternInit object to URLPattern result.
-    return adoptRef(*new URLPattern(WTFMove(processedInit)));
-}
+    Ref result = adoptRef(*new URLPattern());
```
nit: `()` not needed.

cdumez commented on Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp:
```diff
@@ -0,0 +1,69 @@
+
+URLPatternComponent::URLPatternComponent(String&& patternString, JSC::Strong<JSC::RegExp>&& regex, Vector<String>&& groupNameList, bool hasRegexpGroups)
+    : m_patternString(WTFMove(patternString))
+    , m_regularExpression(regex)
```
`WTFMove(regex)`

jelee53 committed:
Add URLPatternComponent class
https://bugs.webkit.org/show_bug.cgi?id=283073
rdar://139833314

Reviewed by Sihui Liu.

Add preliminary support for URLPatternComponent, see https://urlpattern.spec.whatwg.org/#component

* LayoutTests/fast/urlpattern/urlpattern-component-expected.txt:
* LayoutTests/fast/urlpattern/urlpattern-component.html:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.sharedworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/urlpattern/urlpattern.https.any.worker-expected.txt:
* Source/WTF/wtf/URL.cpp:
(WTF::URL::isIPv6Address):
(WTF::isIPv6Address): Deleted.
* Source/WTF/wtf/URL.h:
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::CodePoints::codePointAt const):
* Source/WebCore/Modules/url-pattern/URLPattern.cpp:
(WebCore::processBaseURLString):
(WebCore::URLPattern::create):
(WebCore::URLPattern::compileAllComponents):
(WebCore::escapePatternStringForCharacters): Deleted.
(WebCore::escapePatternString): Deleted.
(WebCore::URLPattern::URLPattern): Deleted.
* Source/WebCore/Modules/url-pattern/URLPattern.h:
* Source/WebCore/Modules/url-pattern/URLPattern.idl:
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp:
(WebCore::canonicalizeOpaquePathname):
(WebCore::canonicalizePathname):
(WebCore::callEncodingCallback):
* Source/WebCore/Modules/url-pattern/URLPatternCanonical.h:
* Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp: Added.
(WebCore::URLPatternUtilities::URLPatternComponent::URLPatternComponent):
(WebCore::URLPatternUtilities::URLPatternComponent::compile):
* Source/WebCore/Modules/url-pattern/URLPatternComponent.h: Copied from Source/WebCore/Modules/url-pattern/URLPatternCanonical.h.
(WebCore::URLPatternUtilities::URLPatternComponent::patternString const):
* Source/WebCore/Modules/url-pattern/URLPatternParser.cpp:
(WebCore::URLPatternUtilities::URLPatternParser::performParse):
(WebCore::URLPatternUtilities::URLPatternParser::addPart):
(WebCore::URLPatternUtilities::URLPatternParser::parse):
(WebCore::URLPatternUtilities::escapeRegexString):
(WebCore::URLPatternUtilities::convertModifierToString):
(WebCore::URLPatternUtilities::generateRegexAndNameList):
(WebCore::URLPatternUtilities::generatePatternString):
(WebCore::URLPatternUtilities::escapePatternStringForCharacters):
(WebCore::URLPatternUtilities::escapePatternString):
(WebCore::URLPatternUtilities::isValidNameCodepoint):
(): Deleted.
* Source/WebCore/Modules/url-pattern/URLPatternParser.h:
(WebCore::URLPatternUtilities::URLPatternParser::takePartList):
(WebCore::URLPatternUtilities::URLPatternParser::getPartList const): Deleted.
* Source/WebCore/Modules/url-pattern/URLPatternTokenizer.cpp:
(WebCore::URLPatternUtilities::Tokenizer::tokenize):
(WebCore::URLPatternUtilities::isValidNameCodepoint): Deleted.
* Source/WebCore/Modules/webauthn/PublicKeyCredential.cpp:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/286965@main
